src/authenticationinapp/authenticationinapplistener.cpp

```diff
@@ -0,0 +1,592 @@
+            processRequestFailure(error, data);
+          });
+
+  connect(request, &NetworkRequest::requestCompleted,
+          [this](const QByteArray& data) {
+            logger.log() << "Oauth code creation completed" << data;
+
+            QJsonDocument json = QJsonDocument::fromJson(data);
+            if (json.isNull()) {
+              MozillaVPN::instance()->errorHandle(
+                  ErrorHandler::AuthenticationError);
+              return;
+            }
+
+            QJsonObject obj = json.object();
+            QJsonValue code = obj.value("code");
+            if (!code.isString()) {
+              logger.log() << "FxA Authz: code not found";
+              MozillaVPN::instance()->errorHandle(
+                  ErrorHandler::AuthenticationError);
+              return;
+            }
+            QJsonValue state = obj.value("state");
+            if (!state.isString()) {
+              logger.log() << "FxA Authz: state not found";
+              MozillaVPN::instance()->errorHandle(
+                  ErrorHandler::AuthenticationError);
+              return;
+            }
+            QJsonValue redirect = obj.value("redirect");
+            if (!redirect.isString()) {
+              logger.log() << "FxA Authz: redirect not found";
+              MozillaVPN::instance()->errorHandle(
+                  ErrorHandler::AuthenticationError);
+              return;
+            }
+
+            NetworkRequest* request =
+                NetworkRequest::createForGetUrl(this, redirect.toString(), 200);
+
+            connect(request, &NetworkRequest::requestFailed,
+                    [this](QNetworkReply::NetworkError error,
+                           const QByteArray& data) {
+                      logger.log()
+                          << "Failed to fetch the final redirect data" << error;
+                      processRequestFailure(error, data);
+                    });
+
+            connect(request, &NetworkRequest::requestHeaderReceived,
+                    [this](NetworkRequest* request) {
+                      // On a 200 response, we receive the OAuth code from the
+                      // query string
+                      QString code =
+                          QUrlQuery(request->url()).queryItemValue("code");
+                      if (code.isEmpty()) {
+                        logger.log() << "Code not received!";
+                        MozillaVPN::instance()->errorHandle(
+                            ErrorHandler::AuthenticationError);
+                        emit failed(ErrorHandler::AuthenticationError);
+                        return;
+                      }
+
+                      emit completed(code);
+                    });
+          });
+}
+
+void AuthenticationInAppListener::processErrorCode(int errorCode) {
+  AuthenticationInApp* aip = AuthenticationInApp::instance();
+  Q_ASSERT(aip);
+
+  switch (errorCode) {
+    case 101:  // Account already exists
```

Comment:
I dislike seeing naked integers hanging around in the code, I feel like these should be an `enum`.

Reply:
I agree. Would be OK-ish if I write a URL of where the numbers are defined?

Reply:
That would certainly be better